### DISCUSSION TITLE
Writing Hard Disk Slices

### DISCUSSION THREAD

**Discussion by hubertushirsch** [wwarthen/RomWBW#457] [Show and tell]
Referring to section 7.2.4 in the RomWBW User Guide, Writing Hard Disk Slices, I would like to add the following.

This procedure also works under Windows.
I have installed the MSYS2 development environment (https://www.msys2.org/) on my Windows computer.
There I can work in an MSYS shell like under Linux and also have the 'dd' command available.
The media that Windows knows natively are mapped in the MSYS shell as /dev/sd{x} and can be accessed accordingly.

The medium in my card reader, which I am using here as an example, is mapped in MSYS as /dev/sdg.
The card was written with the hd1k combo image, which looks like this in MSYS fdisk:
```
admin@MyPC MSYS .../Binary # fdisk /dev/sdg

Welcome to fdisk (util-linux 2.35.2).
Changes will remain in memory only, until you decide to write them.
Be careful before using the write command.

Command (m for help): p
Disk /dev/sdg: 971.59 MiB, 1018773504 bytes, 1989792 sectors
Units: sectors of 1 * 512 = 512 bytes
Sector size (logical/physical): 512 bytes / 512 bytes
I/O size (minimum/optimal): 512 bytes / 512 bytes
Disklabel type: dos
Disk identifier: 0x9921862d

Device     Boot   Start     End Sectors  Size Id Type
/dev/sdg1          2048 1050623 1048576  512M 2e unknown
/dev/sdg2       1050624 1837055  786432  384M  6 FAT16

Command (m for help): q
```
To write a specific slice with an image, I proceed in the same way as described in 7.2.4, with one exception:
I do not address the entire device (/dev/sdg) as the outfile, but rather the RomWBW partition /dev/sdg1.
The advantage is:
- When calculating the offset, I do not have to take into account the size of the prefix, which may be a different size than 1MB when partitioning manually.

Example 2 from the user guide then looks like this for me:
```
user@MyPC MSYS .../Binary # dd if=hd1k_games.img of=/dev/sdg1 seek=48 bs=1M
```
The seek=48 skips the first 48MB of the partition into which we are writing the image.
48 is calculated as (slice number * 8), where 8 is the size of a slice. 

I have found that I do not need elevated rights to access the partition, no need to start the
MSYS shell with 'Run as Administrator'. I can work as a normal, unprivileged user.
To access the entire device (/dev/sdg) I must have elevated rights.
I cannot check on a native Linux system whether an unprivileged user can then do without 'sudo'.
```
user@MyPC MSYS .../Binary $ dd if=/dev/sdg of=/dev/null bs=512 count=1
dd: failed to open '/dev/sdg': Permission denied

user@MyPC MSYS .../Binary $ dd if=/dev/sdg1 of=/dev/null bs=512 count=1
1+0 records in
1+0 records out
512 bytes copied, 0.0043061 s, 119 kB/s
```
Hubert

**Comment by wwarthen** [comment]
Hi @hubertushirsch,

Thank you for bringing this up.  I think you have made two very good points that should be incorporated into the User Guide.

First is that you do not need to use Linux/MaxOS to take advantage of `dd` and manipulate slices individually.  In addition to MSYS, I believe this can be done with CygWin.  Even easier is [dd for Windows](http://www.chrysocome.net/dd).  This is a native Windows version of `dd` and is essentially the same as dd under Linux.

Second is that you can point directly to the RomWBW hd1k partition do do the slice manipulations.  This is much safer and easier than pointing to the full faw disk.

I am out of town this week, but will work on updating the documentation sometime next week.

Thanks! Wayne

**Reply by hubertushirsch** [reply]
Hi Wayne,

you are absolutely right. There is 'dd' for Windows in various projects. Another one is Gnuwin32.

MSYS2 is not Cygwin, thats different projects
https://www.msys2.org/wiki/How-does-MSYS2-differ-from-Cygwin/

In MSYS2, 'dd' and many other UNIX like tools are binaries (EXE Files), that run standalone in
a Windows CMD (Dos-Box). Most of this you can found in the <msys2_install_path>/usr/bin.

But in an MSYS shell (a 'bash') I have a complete Linux-typical environment with all the environment variables that I need to use the compiler toolchains.
I can also automate a lot of things with shell scripts.
Which I couldn't do effectively if I called the programs individually in a Windows shell (CMD, PowerShell).

For example, my backup script. In my project, I only need to back up the first 640kB from each slice, as they are just floppy disk images.
```
#!/bin/sh
# Sicherung von RomWBW RCBD UDOS CF- und SD-Karten
# 128MB - 15 Slices
size=8192
device=/dev/sdg1
base=backup.128m
dd if=$device of=${base}.slice00 bs=1k count=640 skip=$(($size*0))
dd if=$device of=${base}.slice01 bs=1k count=640 skip=$(($size*1))
dd if=$device of=${base}.slice02 bs=1k count=640 skip=$(($size*2))
dd if=$device of=${base}.slice03 bs=1k count=640 skip=$(($size*3))
dd if=$device of=${base}.slice04 bs=1k count=640 skip=$(($size*4))
dd if=$device of=${base}.slice05 bs=1k count=640 skip=$(($size*5))
#dd if=$device of=${base}.slice06 bs=1k count=640 skip=$(($size*6))
#dd if=$device of=${base}.slice07 bs=1k count=640 skip=$(($size*7))
#dd if=$device of=${base}.slice08 bs=1k count=640 skip=$(($size*8))
#dd if=$device of=${base}.slice09 bs=1k count=640 skip=$(($size*9))
# Slice 10 (0a): UDOS Loader Z180
dd if=$device of=${base}.slice10 bs=1k count=640 skip=$(($size*10))
# Slice 11 (0b): UDOS Loader Z80
dd if=$device of=${base}.slice11 bs=1k count=640 skip=$(($size*11))
# Slice 12 (0c): Default System Disk
dd if=$device of=${base}.slice12 bs=1k count=640 skip=$(($size*12))
#dd if=$device of=${base}.slice13 bs=1k count=640 skip=$(($size*13))
#dd if=$device of=${base}.slice14 bs=1k count=640 skip=$(($size*14))
# End of slices
```
Try doing that without a script and without any errors.
 
Good days
Hubert

**Reply by wwarthen** [reply]
Hi @hubertushirsch,

Good information about MSYS.  You have done a great job putting together some automation related to the slices.  Well done.

I am not sure how far I want to go into a specific OS/tool solution in the User Guide.  Everyone seems to have their favorite way to do things.  I do think you have made some points that do need to be included and will do so.

Note that I am out of town this week, so I won't get around to making changes until next week.

Thanks, Wayne

**Comment by hubertushirsch** [comment]
Hello Wayne,

I don't expect you to include any scripts from me in the documentation.

I just wanted to explain why I prefer to work in a Linux-like environment rather than with singular tools under Windows.
I don't run Linux on hardware, only in a virtual machine.
My multi-card reader doesn't work there. The MSYS environment is the ideal replacement for me.

So the message is: **dd** is also available under Windows, the user has a choice of several solutions.

Just for your information:
I tried the **dd** from your link http://www.chrysocome.net/dd.
Since it is a native Windows program, you have to work with the Windows device identifiers when accessing devices.
After a while I found out which device my CF card is from the perspective of Windows.
These three notations access my RomWBW partition:
```
E:\Service\Devel>
E:\Service\Devel>dd if=\\.\Volume{1087ffd4-816d-11ef-bc4c-d35bc053ef81} of=test1.img bs=1k count=640 skip=81920
rawwrite dd for windows version 1.0beta1 WIN64.
Written by John Newbigin <jnewbigin@chrysocome.net>
This program is covered by terms of the GPL Version 2.

skip to 83886080
640+0 records in
640+0 records out

E:\Service\Devel>dd if=\\?\Device\HarddiskVolume10 of=test2.img bs=1k count=640 skip=81920
rawwrite dd for windows version 1.0beta1 WIN64.
Written by John Newbigin <jnewbigin@chrysocome.net>
This program is covered by terms of the GPL Version 2.

skip to 83886080
640+0 records in
640+0 records out

E:\Service\Devel>dd if=\\.\g: of=test3.img bs=1k count=640 skip=81920
rawwrite dd for windows version 1.0beta1 WIN64.
Written by John Newbigin <jnewbigin@chrysocome.net>
This program is covered by terms of the GPL Version 2.

skip to 83886080
640+0 records in
640+0 records out
```
Hubert

**Comment by wwarthen** [comment]
I have attempted to address this in commit b032694e0139d66b740b215dbd44a441ddc97822.

Thanks, Wayne
